### PR TITLE
New methods for books and borrow records

### PR DESCRIPTION
**Pull request by tinwech**
- Add put and delete method to books endpoints to modify books by their IDs
- Add records endpoints for getting borrow records

### CODE REVIEW

File: domain/book.go
Comment (alan890104) [inline domain/book.go:1]: @tinwech You should check all of your interface provides default implementation in repositories. The project is currently not buildable.

Take `BookRepository` as example, `DeleteById(ctx context.Context, id string) error` should be added to `book_repository.go` with default functions.

```go
func (r *bookRepository) DeleteById(ctx context.Context, id string) error {
	return nil
}
```

File: api/restful/openapi.yaml
Comment (alan890104) [inline api/restful/openapi.yaml:1]: For clearly view the api in openapi editor, please add tags to categorize them

**Review by alan890104 — CHANGES_REQUESTED** [review]

**Comment by tinwech** [comment]
Fixed build issue and added tags to API doc.

**Comment by tinwech** [comment]
- update API doc
- fix repository query bug
- finish implementation of book/record usecase and repository

@alan890104 pls help review the code, thanks!

File: api/restful/routes/book_route.go [hunk 1]
```diff
@@ -9,7 +9,9 @@ func RegisterBookRoutes(engine *gin.Engine, bookController *controller.BookContr
 	r := engine.Group("/api/v1/books")
 	r.GET("", bookController.GetBook)
 	r.GET("/:bookID", bookController.GetBookByID)
+	r.PUT("/:bookID", bookController.UpdateBookByID)
+	r.DELETE("/:bookID", bookController.DeleteBookByID)
 	r.POST("", bookController.CreateBook)
-	r.PUT("/borrow/:bookID", bookController.BorrowBook)
-	r.PUT("/return/:bookID", bookController.ReturnBook)
+	r.POST("/borrow", bookController.BorrowBook)
```
Comment (alan890104) [inline api/restful/routes/book_route.go:15]: Have you tested your API after implementing? I notice some inconsistencies between the `book_route.go` file and the `openapi.yaml` specification. Please **ensure code consistency and adherence to the API spec**.

First and foremost, remember that an API should not be mistaken for a proxy to a database; it should distinctly reflect its intended purpose. Secondly, ensure that it adheres to the [RESTful standards](https://aws.amazon.com/tw/what-is/restful-api/). As an illustration, consider actions like 'borrow' and 'return,' which are modifications and, therefore, should be implemented as 'PUT' actions.

File: usecase/record_usecase.go [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
+func (us *recordUsecase) GetLatestByBookId(ctx context.Context, bookId string) (*domain.Record, error) {
+	return us.recordrepo.GetLatestByBookId(ctx, bookId)
+}
+
+func (us *recordUsecase) Store(ctx context.Context, record *domain.StoreRecordRequest) error {
+	return us.recordrepo.Store(ctx, &domain.Record{
+		UserId:    record.UserId,
+		BookId:    record.BookId,
+		StartDate: time.Now().UTC().Format(dateFormat),
```
Comment (alan890104) [inline usecase/record_usecase.go:48]: In this repository, we've initialized the time configuration in `bootstrap`. You might want to consider utilizing `time.Local` in place of UTC.

File: usecase/record_usecase.go [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+	})
+}
+
+func (us *recordUsecase) UpdateEndDateById(ctx context.Context, id string) error {
+	endDate := time.Now().UTC().Format(dateFormat)
```
Comment (alan890104) [inline usecase/record_usecase.go:53]: Same as above

**Review by alan890104 — CHANGES_REQUESTED** [review]

File: api/restful/routes/book_route.go (hunk 1 above)
Comment (tinwech) [reply api/restful/routes/book_route.go:15]: I forgot to discuss the 'borrow' and 'return' actions with you in our last meeting. I modified their APIs to include both book ID and user ID. I am confused about the best API design as these actions modify the book resource and create or modify a record. Which of the following is the best practice in this case? The first design is currently in use.

1. `POST /books/borrow` (bookId and userId in the request body)
2. `PUT /books/borrow/{bookId}?userId={userId}`
3. `PUT /books/borrow/{bookId}` (userId will be included in the JWT payload, not implemented yet though)

BTW, the APIs were manually tested and worked well, should I write unit/integration tests?

Comment (alan890104) [reply api/restful/routes/book_route.go:15]: If there is no payload (user ID will be parsed from the JWT payload), `PUT /books/borrow/{bookId}` would be a better option. A query string (i.e., the string after the question mark in the URL) represents a filter and may only appear for pagination in the `GET` method. (Ofcourse there are some exceptions)

**Pull request merged**